lib/ZumoOta/src/BootloaderCom.cpp

```diff
@@ -307,24 +330,246 @@ class FwProgCmdProvider : public CmdProvider {
+        m_firmwareFile(),
+        m_count(0U),
+        m_firmwareSize(0U),
+        m_max_buffers(0),
+        m_currWriteMemAddr(0x0000)
+     {}
+
+    /**
+ * @brief Resets the firmware file and prepares it for reading.
+ * 
+ * This function closes the currently opened firmware file, if any,
+ * and then attempts to open the firmware file specified by m_fileName
+ * in read mode. It retrieves the size of the firmware file and calculates
+ * the maximum number of buffers based on the file size divided by 128.
+ * @return True if the firmware file is successfully opened, false otherwise.
+ */
+    virtual bool reset()
+    { 
+        /*Check if the file is already open.*/
+        if (m_firmwareFile)
+        {
+            m_firmwareFile.close();
+        }
+
+        /*If the file is not yet open, open it.*/
+        m_firmwareFile = LittleFS.open(m_fileName, "r");
+        /*Retrieves the size of the firmware file and calculates the maximum number of buffers.*/
+        m_firmwareSize = m_firmwareFile.size();
+        LOG_INFO("Size of the file = %d", m_firmwareSize);
+        m_max_buffers = m_firmwareFile.size() / FLASH_BLOCK_LENGTH;
+        if((m_firmwareFile.size() % FLASH_BLOCK_LENGTH) == 0)
+        {
+            /*m_max_buffers doesn't change*/
+        }
+        else
+        {
+            /*Incrementing m_max_buffers by 1 to accommodate the remainder of the file.*/
+            m_max_buffers += 1;
+        }
+        LOG_INFO("m_max_buffers = %d", m_max_buffers);
+        return false != m_firmwareFile;
+    }
+
+    /**
+     * @brief Retrieves the next command and the next response.
+     *
+     * This method reads bytes from the target device, compares them with
+     * the expected bytes, and provides the next command and response accordingly.
+     *
+     * @param [out] cmd Reference to a pointer to the next command.
+     * @param [out] rsp Reference to a pointer to the next response.
+     * @return True if there is a next command and response, false otherwise.
+     */
+    virtual bool next(const CommandInfo*& cmd, const ResponseInfo*& rsp) override
+    {
+        if(m_count < m_max_buffers)
+        {
+            /*Keep original command.*/
+            if( m_nextCmd == CCMD_SET_ADDR)
+            {
+                rsp = &m_responses[m_nextCmd];
+    
+                /*Update the command with required content to write in the flash memory.*/
+                if(true == updateSetAdressCmd(m_updatedCmd))
+                {
+                    cmd = &m_updatedCmd;
+                    return true;
+                }
+            }
+            else if (m_nextCmd == CCMD_READ_BLOCK)
+            {
+                /*Increment the count of processed buffers.*/
+                m_count++;
+                cmd =&m_cmds[m_nextCmd];
+                /*Create a static ResponseInfo pointer to hold temporary response information.*/
+                static ResponseInfo tempResponse[FLASH_BLOCK_LENGTH];

```

Comment:
An array with 512 ResponseInfos ??

Please stop adding statics into methods. use class members.

Reply:
FLASH_BLOCK_LENGTH is 128Bytes. It will be enough to store data.